### PR TITLE
Use years throughout the app based on the APD summary selection

### LINKED ISSUE #462: [Program Overview Section] Federal Fiscal Year back-end work
Allow the selection of FFYs within the program overview section drive all of the tables and content through out the rest of the APD.

### This task is done when...
 - [ ] FFY selection is available within the Overview section
 - [ ] FFYs selected within the Overview section drive all subsequent tables within the APD.

### LINKED ISSUE #630: Zero out Non Personnel Costs in the Activity Section
[Bug] Non Personnel Costs in the activity section still have a default $100 attached to it. This should be zero'd out. Workaround would be user manually zero-ing the number out or updating it when they go to the page.

image

### This task is done when...
 - [ ] Default values in non personnel costs are $0
 - [ ] Someone else has verified the change and it didn't break any of the other tables.

### LINKED ISSUE #631: FFY Selection in Program Overview does not affect cost allocation or summary budget tables
FFY selection in program overview is not affecting the FFYs that are displayed in the Cost Allocation section or the budget tables.

image

image

image

### This task is done when...
 - [ ] FFY selection in the program overview updates the Cost Allocation section in the activities
 - [ ] FFY selection in the program overview updates the Budget Tables
- [ ] Someone else has verified that this changes have been made.

### LINKED ISSUE #632: Program Overview - FFY's displayed should be dynamically generated
The Program Overview in this app currently shows FFY 2018, FFY 2019 and FFY 2020 which is hard coded  into the app. Instead, this should be dynamically generated based on the FFY we are in.

image

HITECH funding can be approved for the current FFY + 2 years out. 

HITECH funding is proposed to end in 2023 (pending final rule + limits on what we can support beyond 2021). Maybe not for MVP, but we would not want to display years where funding is not available (though a work around would be for CMS to deny any funding beyond a requested date).

FFY runs October 1, 20XX  - September 30, (20XX+1). 

Example FFY 2019 will run from October 1, 2018 through September 30, 2019.

### This task is done when...
 - [ ] Help text updated (if needed)
 - [ ] FFYs are dynamically generated and working correctly with the rest of the app
- [ ] Product has approved the experience.

### LINKED ISSUE #641: Make years consistent throughout the app
The contents of all the other hunks of the APD should reflect the years selected in the program overview section.

### PR DESCRIPTION
**Pull request by mgwalker**
Describe the pull request here, including any supplemental information needed to understand it.

Closes #462
Closes #630
Closes #631
Closes #632
Closes #641

### This pull request changes...
- no hidden 2020 expenses!
- if you DESELECT a year for which data has been entered, that data will be lost
  - need a separate issue to discuss/design the right solution - could just hide that data, or prompt the user to confirm that they want to destroy it
  - if we just hide it, what happens when the APD is saved?

### This pull request is ready to merge when...
- [ ] Tests have been updated (and all tests are passing)
- [ ] This code has been reviewed by someone other than the original author

### This feature is done when...
- [ ] Product has approved the experience

### CODE REVIEW

File: web/src/actions/apd.js [hunk 1]
```diff
@@ -9,14 +9,23 @@ export const SAVE_APD_REQUEST = 'SAVE_APD_REQUEST';
...
 export const receiveApd = data => ({ type: GET_APD_SUCCESS, data });
 export const failApd = error => ({ type: GET_APD_FAILURE, error });
 
 export const addApdKeyPerson = () => ({ type: ADD_APD_KEY_PERSON });
 export const removeApdKeyPerson = id => ({ type: REMOVE_APD_KEY_PERSON, id });
-export const updateApd = updates => ({ type: UPDATE_APD, updates });
+export const updateApd = updates => dispatch => {
+  dispatch({ type: UPDATE_APD, updates });
+  if (updates.years) {
+    dispatch(updateBudget());
```
Comment (mgwalker) [inline web/src/actions/apd.js]: Need to update the budget if the years for the APD were changed.

**Comment by jeromeleecms** [comment]
@mgwalker (some comments below do not directly apply to the functionality, so let me know if you want to spin up separate issues on these)

- Core function appears to be working as advertised.

- Not sure if this is related, but there is no default activity listed anymore. When I add an activity, I can't type in the box (though I can if I add a second activity).  

image

- Cost Allocation doesn't seem to be translating into the budget section (seems to be fixed in master). 

- Allows user to not have any FFY selected (maybe for the MVP this is OK, but a warning would be nice)

- Allows user to select non consecutive years (maybe for the MVP this is OK, but to your original post, if they realize that they did this and then deselect it, they'll lose all the data. Maybe better to hide it in case they need to reference it and then they can deselect it. But worth a discussion)

- Will this apply to the Incentive Payments by FFY Quarter and the Overall Budget by FFY Quarter as well (since these are not built yet)?

**Comment by mgwalker** [comment]
@jeromeleecms:

* >Not sure if this is related, but there is no default activity listed anymore. When I add an activity, I can't type in the box (though I can if I add a second activity).

  I tried to fix that, but I guess I didn't.  I'll go back in and see if I can figure out what's going on.  It's a quirk of the data, though, partly a manifestation of the fact that we don't have a mechanism for creating new APDs.  Right now, the APD you edit was hand-jammed into the database.  Who knows what kind of stuff I messed up?  😬 This one feels like a blocking bug.

* >Cost Allocation doesn't seem to be translating into the budget section (seems to be fixed in master).

  Can you say more about you're (not) seeing?  When I add expenses and tweak the cost allocation, the summary budget table updates accordingly.  We can hop on video if it'd be easier to show than tell.  This one also feels like a blocking bug.

* >Allows user to not have any FFY selected (maybe for the MVP this is OK, but a warning would be nice)

  > Allows user to select non consecutive years (maybe for the MVP this is OK, but to your original post, if they realize that they did this and then deselect it, they'll lose all the data. Maybe better to hide it in case they need to reference it and then they can deselect it. But worth a discussion)

  Definitely worth a discussion.  I'd lean towards addressing these in a separate issue because I think they might take some time to work through how we want to do it and unless this is fundamentally broken, I'd not want to hold it up.

**Comment by jeromeleecms** [comment]
@mgwalker Working on recreating the cost allocation bug and found another one. In the default Summary Budget Table, this shows up (it goes away when you select different FFY's though):

image

Will get back to you on the cost allocation item in a sec.

**Comment by mgwalker** [comment]
So both the default activity bug and the weird budget table appearance bug are because there's no APD saved in the database for you.  The app doesn't know what to do so you start with an incomplete APD object in state and that causes other things to break downstream.

This shouldn't be too much effort to fix, and this seems like a valid condition - where a user is creating a brand new APD, it won't exist in the database yet.  We should be able to do that!  😅

**Comment by jeromeleecms** [comment]
@mgwalker Nevermind about the cost allocation bug -- it appears I was imagining things.. or trying to do mental math, which resulted in me imagining things, which made me want to do more mental math... which...

**Comment by mgwalker** [comment]
Default activity and initial budget summary view bugs are fixed.  🎉

**Comment by jeromeleecms** [comment]
Per Standup, this request is tied to #462, #632, #630, #631

**Pull request merged**